pkg/autodiscovery/providers/prometheus_pods.go

```diff
@@ -0,0 +1,150 @@
+import (
+	"encoding/json"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/providers/names"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/kubelet"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// PrometheusPodsConfigProvider implements the ConfigProvider interface for prometheus pods.
+type PrometheusPodsConfigProvider struct {
+	kubelet kubelet.KubeUtilInterface
+	checks  []*PrometheusCheck
+}
+
+// NewPrometheusPodsConfigProvider returns a new Prometheus ConfigProvider connected to kubelet.
+// Connectivity is not checked at this stage to allow for retries, Collect will do it.
+func NewPrometheusPodsConfigProvider(config config.ConfigurationProviders) (ConfigProvider, error) {
+	p := &PrometheusPodsConfigProvider{}
+	err := p.setupConfigs()
+	return p, err
+}
+
+// setupConfigs reads and initializes the checks from the configuration
+// It defines a default openmetrics instances with default AD if the checks configuration is empty
+func (p *PrometheusPodsConfigProvider) setupConfigs() error {
+	checks := []*PrometheusCheck{}
+	if err := config.Datadog.UnmarshalKey("prometheus_scrape.checks", &checks); err != nil {
+		return err
+	}
+
+	if len(checks) == 0 {
+		log.Info("The 'prometheus_scrape.checks' configuration is empty, a default openmetrics check configuration will be used")
+		p.checks = []*PrometheusCheck{defaultCheck}
+		return nil
+	}
+
+	p.checks = checks
+	for _, check := range p.checks {
+		check.init()
+	}
+
+	return nil
+}
+
+// String returns a string representation of the PrometheusPodsConfigProvider
+func (p *PrometheusPodsConfigProvider) String() string {
+	return names.Prometheus
+}
+
+// Collect retrieves templates from the kubelet's podlist, builds config objects and returns them
+func (p *PrometheusPodsConfigProvider) Collect() ([]integration.Config, error) {
+	var err error
+	if p.kubelet == nil {
+		p.kubelet, err = kubelet.GetKubeUtil()
+		if err != nil {
+			return []integration.Config{}, err
+		}
+	}
+
+	pods, err := p.kubelet.GetLocalPodList()
+	if err != nil {
+		return []integration.Config{}, err
+	}
+
+	return p.parsePodlist(pods), nil
+}
+
+// IsUpToDate always return false to poll new data from kubelet
+func (p *PrometheusPodsConfigProvider) IsUpToDate() (bool, error) {
+	return false, nil
+}
+
+// parsePodlist searches for pods that match the AD configuration
+func (p *PrometheusPodsConfigProvider) parsePodlist(podlist []*kubelet.Pod) []integration.Config {
+	configs := []integration.Config{}
+	for _, pod := range podlist {
+		for _, check := range p.checks {
+			configs = append(configs, check.configsForPod(pod)...)
+		}
+	}
+	return configs
+}
+
+// configsForPod returns the openmetrics configurations for a given pod if it matches the AD configuration
+func (pc *PrometheusCheck) configsForPod(pod *kubelet.Pod) []integration.Config {
+	configs := []integration.Config{}
```

Comment:
Minor nit - you could leave this as `nil` and return as `nil` if no `append` happens, to spare the allocation.

Reply:
do you mean: `var config []integration.Config` ?

Reply:
Yes.